File: packages/maps-launcher/ios/Plugin/MapsLauncherPlugin.swift
import Foundation
import Capacitor

@objc(MapsLauncherPlugin)
public class MapsLauncherPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MapsLauncherPlugin"
    public let jsName = "MapsLauncher"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getAvailableApps", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getDefaultApp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "navigate", returnType: CAPPluginReturnPromise)
    ]
    public static let tag = "MapsLauncherPlugin"

    private var implementation: MapsLauncher?

    override public func load() {
        self.implementation = MapsLauncher(plugin: self)
    }

    @objc func getAvailableApps(_ call: CAPPluginCall) {
        implementation?.getAvailableApps { result, error in
            if let error = error {
                self.rejectCall(call, error)
                return
            }
            self.resolveCall(call, result)
        }
    }

    @objc func getDefaultApp(_ call: CAPPluginCall) {
        rejectCallAsUnimplemented(call)
    }

    @objc func navigate(_ call: CAPPluginCall) {
        do {
            let options = try NavigateOptions(call)
            implementation?.navigate(options) { error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call)
            }
        } catch {
            self.rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", MapsLauncherPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func rejectCallAsUnimplemented(_ call: CAPPluginCall) {
        call.unimplemented("This method is not available on this platform.")
    }

    private func resolveCall(_ call: CAPPluginCall) {
        call.resolve()
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}

File: packages/maps-launcher/android/src/main/java/io/capawesome/capacitorjs/plugins/mapslauncher/MapsLauncherPlugin.java
package io.capawesome.capacitorjs.plugins.mapslauncher;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.mapslauncher.classes.CustomException;
import io.capawesome.capacitorjs.plugins.mapslauncher.classes.options.NavigateOptions;
import io.capawesome.capacitorjs.plugins.mapslauncher.classes.results.GetAvailableAppsResult;
import io.capawesome.capacitorjs.plugins.mapslauncher.classes.results.GetDefaultAppResult;
import io.capawesome.capacitorjs.plugins.mapslauncher.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.mapslauncher.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.mapslauncher.interfaces.Result;

@CapacitorPlugin(name = "MapsLauncher")
public class MapsLauncherPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "MapsLauncherPlugin";

    private MapsLauncher implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new MapsLauncher(this);
    }

    @PluginMethod
    public void getAvailableApps(PluginCall call) {
        try {
            NonEmptyResultCallback<GetAvailableAppsResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull GetAvailableAppsResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.getAvailableApps(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void getDefaultApp(PluginCall call) {
        try {
            NonEmptyResultCallback<GetDefaultAppResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull GetDefaultAppResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.getDefaultApp(callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void navigate(PluginCall call) {
        try {
            NavigateOptions options = new NavigateOptions(call);
            EmptyCallback callback = new EmptyCallback() {
                @Override
                public void success() {
                    resolveCall(call);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.navigate(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}
